crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -279,66 +295,343 @@ where
         Ok(())
     }
 
+    /// Look up the IBC denomination from a IbcToken.
+    pub fn query_ibc_denom(
+        &self,
+        token: impl AsRef<str>,
+        owner: Option<&Address>,
+    ) -> Result<String> {
+        let hash = match Address::decode(token.as_ref()) {
+            Ok(Address::Internal(InternalAddress::IbcToken(hash))) => {
+                hash.to_string()
+            }
+            _ => return Ok(token.as_ref().to_string()),
+        };
+
+        if let Some(owner) = owner {
+            let ibc_trace_key = ibc_trace_key(owner.to_string(), &hash);
+            if let Some(ibc_denom) = self.ctx.read_pre(&ibc_trace_key)? {
+                return Ok(ibc_denom);
+            }
+        }
+
+        // No owner is specified or the owner doesn't have the token
+        let ibc_denom_prefix = ibc_trace_key_prefix(None);
+        let ibc_denoms = self.ctx.iter_prefix(&ibc_denom_prefix)?;
+        for (key, ibc_denom, _gas) in ibc_denoms {
+            if let Some((_, token_hash)) =
+                is_ibc_trace_key(&Key::parse(key).into_storage_result()?)
+            {
+                if token_hash == hash {
+                    return String::try_from_slice(&ibc_denom[..])
+                        .into_storage_result()
+                        .map_err(Error::NativeVpError);
+                }
+            }
+        }
+
+        Ok(token.as_ref().to_string())
+    }
+
+    // Apply the given send packet to the changed balances structure
+    fn apply_send_packet(
+        &self,
+        acc: &mut ChangedBalances,
```

Comment:
Is there a reason we take `acc` as a `mut` argument instead of simply returning a `ChangedBalances` which can be added or something? In general, I dislike the pattern of taking arguments as `mut` which are effectively return types.

Reply:
There's no reason. Returning `ChangedBalances` also works. Will apply this change.